### PR TITLE
Core: Temporary unavailable replicas; Fix #1550

### LINKED ISSUE #1550: Mark replicas as temporarily unavailable
Motivation
----------

> This comes up again for LRZ as we have a broken server. What is possible now, if we have a list of files to be marked temporarily unavailable? I am no longer pushing for rucio to know the pool info - a list produced by the site is fine, if it can be handled in bulk.
> Files marked unavailable would
> - not show up in list-replicas, also in --deep dataset replica lookup used by panda
> - mark complete datasets as not complete (panda uses complete replicas, to avoid file replica lookup)
> - not trigger recovery from other replicas

The critical part here is what happens with the replication rules touching these replicas. The state-consistent workflow would be, to set them to STUCK, but this would issue an immediate recovery of these replicas. Leaving the rules in OK state breaks somewhat the consistency of the state-machine and could lead to problems if admins forget to set these replicas back to AVAILABLE again. 

Modification
------------
- [ ] CLI command needed to operate over a list of replicas supplied in a txt file
- [ ] Replicas are marked as UNAVAILABLE
- [ ] Collection Replicas are updated respectively
- [ ] Reverse CLI command needed to make replicas and collection replicas AVAILABLE again

### PR DESCRIPTION
**Pull request by bari12**
Core: Temporary unavailable replicas; Fix #1550

### CODE REVIEW

**Comment by bari12** [comment]
This is a big PR, probably will need some refinement. Important thing is to make sure that it doesn't break anything.

**Comment by cserf** [comment]
Typo : `mins_temporary_expiration.py` --> `minos_temporary_expiration.py`

File: lib/rucio/core/replica.py [hunk 1]
```diff
@@ -2320,3 +2326,193 @@ def mark_unlocked_replicas(rse, bytes, session=None):
...
+                                             state=state, account=account, bytes=None, expires_at=expires_at)
+        new_bad_replica.save(session=session, flush=False)
+    try:
+        session.flush()
+    except IntegrityError as error:
+        raise exception.RucioException(error.args)
+    except DatabaseError as error:
+        raise exception.RucioException(error.args)
+    except FlushError as error:
+        if match('New instance .* with identity key .* conflicts with persistent instance', error.args[0]):
```
Comment (mlassnig) [inline lib/rucio/core/replica.py:2545]: is this oracle specific? what are the equivalent calls for slite/mysql/psql?

File: lib/rucio/core/replica.py [hunk 2]
```diff
@@ -2320,3 +2326,193 @@ def mark_unlocked_replicas(rse, bytes, session=None):
...
+        new_pfn = models.BadPFNs(path=str(pfn), account=account, state=rep_state, reason=reason, expires_at=expires_at)
+        new_pfn.save(session=session, flush=False)
+    try:
+        session.flush()
+    except IntegrityError as error:
+        raise exception.RucioException(error.args)
+    except DatabaseError as error:
+        raise exception.RucioException(error.args)
+    except FlushError as error:
+        if match('New instance .* with identity key .* conflicts with persistent instance', error.args[0]):
```
Comment (mlassnig) [inline lib/rucio/core/replica.py:2623]: same

File: lib/rucio/daemons/badreplicas/minos.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+                            session.commit()  # pylint: disable=no-member
+                    except Exception:
+                        session.rollback()  # pylint: disable=no-member
+                        logging.critical(traceback.format_exc())
+
+        except Exception as error:
+            logging.error(prepend_str + '%s' % (str(error)))
+
+        tottime = time.time() - start_time
+        if once is True:
```
Comment (mlassnig) [inline lib/rucio/daemons/badreplicas/minos.py]: if once:

File: lib/rucio/daemons/badreplicas/minos_temporary_expiration.py [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+        try:
+            # Get list of expired TU replicas
+            expired_replicas = list_expired_temporary_unavailable_replicas(total_workers=heart_beat['nr_threads'] - 1,
+                                                                           worker_number=heart_beat['assign_thread'],
+                                                                           limit=1000)
+
+            replicas = []
+            bad_replicas = []
+            for replica in expired_replicas:
+                print replica
```
Comment (mlassnig) [inline lib/rucio/daemons/badreplicas/minos_temporary_expiration.py]: debug stmt?

File: lib/rucio/daemons/badreplicas/minos_temporary_expiration.py [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+                # Remove the replicas from bad_replicas table in chunks
+                bulk_delete_bad_replicas(chunk, session=session)
+
+            session.commit()
+
+        except Exception as error:
+            logging.error(prepend_str + '%s' % (str(error)))
+
+        tottime = time.time() - start_time
+        if once is True:
```
Comment (mlassnig) [inline lib/rucio/daemons/badreplicas/minos_temporary_expiration.py]: if once

**Review by mlassnig — COMMENTED** [review]
also add to daemons docker?

File: lib/rucio/core/replica.py (hunk 1 above)
Comment (bari12) [reply lib/rucio/core/replica.py:2545]: It's SQLA specific.

File: lib/rucio/daemons/badreplicas/minos_temporary_expiration.py (hunk 4 above)
Comment (bari12) [reply lib/rucio/daemons/badreplicas/minos_temporary_expiration.py]: yes, will remove

File: lib/rucio/core/replica.py [hunk 6]
```diff
@@ -1659,6 +1663,8 @@ def update_replicas_states(replicas, nowait=False, session=None):
                                             broken_rule_id=replica.get('broken_rule_id', None),
                                             broken_message=replica.get('broken_message', None),
                                             nowait=nowait, session=session)
+        elif replica['state'] == ReplicaState.TEMPORARY_UNAVAILABLE:
+            query = query.filter_by(state=ReplicaState.AVAILABLE)
```
Comment (cserf) [inline lib/rucio/core/replica.py:1668]: I don't get this line

File: lib/rucio/core/rule.py [hunk 7]
```diff
@@ -2868,10 +2868,11 @@ def __delete_lock_and_update_replica(lock, purge_replicas=False, nowait=False, s
                 replica.tombstone = replica.accessed_at
             else:
                 replica.tombstone = replica.created_at
-        if lock.state == LockState.REPLICATING and replica.lock_cnt == 0:
-            return True
+            if lock.state == LockState.REPLICATING:
+                replica.state = ReplicaState.UNAVAILABLE
```
Comment (cserf) [inline lib/rucio/core/rule.py:2900]: Why do you set the replica state to 'U' whereas it was not the case before?

File: lib/rucio/core/replica.py (hunk 6 above)
Comment (bari12) [reply lib/rucio/core/replica.py:1668]: This is to make sure that the state transition can only go AVAILABLE -> TU

File: lib/rucio/core/rule.py (hunk 7 above)
Comment (bari12) [reply lib/rucio/core/rule.py:2900]: It's just a re-arrange of the conditions as it looks simpler like this. It was set to UNAVAILABLE before too, it just happened in the lines 2860+2861 before.

**Review by cserf — DISMISSED** [review]

**Comment by cserf** [comment]
Good luck for the merge :laughing:

**Comment by vingar** [comment]
I think the corresponding entry in the documentation is missing for the new daemon.

**Pull request merged**